### DISCUSSION TITLE
problem connecting observable to source

### DISCUSSION THREAD

**Discussion by kingjordan** [state-adapt/state-adapt#77] [Q&A] [answered]
Hello,

I am having a problem connecting an observable to a store source in a demo angular app i'm using to learn state-adapt.

I have a call to a firebase auth service that returns a promise which I convert into an observable.

```javascript
  login$(credentials: Credentials) {
    return from(
      defer(() =>
        signInWithEmailAndPassword(
          this.auth,
          credentials.email,
          credentials.password
        )
      ).pipe(toSource('[AUTH] user login'))
    );
  }
```

Here is my store declaration:
```javascript
  private store = adapt(['app state', initialState, storeAdapter], {
    userLogin: this.login$,
  });
```

When I attempt to use this "action source" login$ there is no "next" available on the object so I cannot use it in my templates.  Am I missing something, should not toSource and New Source product the same object type?

for example ` loginUser$ = new Source<Credentials>('[AUTH] user Login');` has a next that can be used int he templates.  How do I connect the two together or adjust my first one to get back a source type?

Thanks

**Comment by mfp22** [comment]
There are 2 events: Request, and response. 

So I would create either a `Subject` or a `Source` to represent the request. 

```typescript
login$ = new Source<Credentials>('login$');
```

The 2nd event comes from the HTTP observable. In order to make the request, you need the `login$` event and its data.

```typescrtipt
loginSuccess$ = this.login$.pipe(
  exhaustMap(credentials => signInWithEmailAndPassword(
          this.auth,
          credentials.email,
          credentials.password
  )),
  toSource('[AUTH] loginSuccess$'),
);
```

`exhaustMap` ignores clicks that occur while the inner observable is still ongoing, so you don't get multiple. 

I recommend reading the full section on sources in the docs: https://state-adapt.github.io/docs/rxjs#source

**Reply by kingjordan** [reply]
thanks for the follow up Mike, its greatly appreciated.  

If I define my loginSuccess$ as you show here it is never triggered if I call `login$.next($event)` from somewhere else in my app.  I have to place it inside my class constructor and subscribe to it in order to get loginSuccess to execute.  Is this proper?

```javascript
  constructor() {
    this.login$
      .pipe(
        exhaustMap((action) =>
          from(
            signInWithEmailAndPassword(
              this.auth,
              action.payload.email,
              action.payload.password
            )
          )
        ),
        toSource('[AUTH] user authenticated')
      )
      .subscribe((result) => console.log(result));
  }
```

therefore I don't define a loginSuccess but rather just subscribe to login$ to trigger the "EFFECT" of logging in.

**Reply by mfp22** [reply, marked as answer]
Effects should not be separate things. 

The store wants the result of that effect, so the store should reference it directly, like you had:

```typescript
  private store = adapt(['app state', initialState, storeAdapter], {
    userLogin: this.loginSuccess$,
  });
```

Okay but stores just represent state. Who cares? Literally, what is that state used for? Usually it's a specific page, so when you subscribe to that store, _then_ the HTTP requests are triggered, and when you leave the page, the state is emptied so a fresh copy can be created when the page is revisited. It's very similar to TanStack Query, if you've ever used that. 

But when you have something that's truly global and you want it to immediately get set up and never have it torn down, then you can subscribe to the store right there next to it to keep it there:

```typescript
  private store = adapt(['app state', initialState, storeAdapter], {
    userLogin: this. loginSuccess$,
  });
  sub = this.store.subscribe(userState => console.log('userState', userState));
```

Either way, the subscription needs to make it to the observable that represents the thing you want. What you did works, but it's cleaner to avoid imperative code and reference the observable in the store. 

I knew this would be a thing that would trip people up, so I tried to include a few reminders in the docs about it:
![image](https://github.com/state-adapt/state-adapt/assets/13700926/a0cafc7a-8853-4af7-bc6a-45a72a69df81)

But I noticed the docs talk about stores and subscriptions a few times, but they may be missing something about sources and subscriptions. I'll update them next time I release a version.

**Comment by kingjordan** [comment]
Once again, thanks for the help.
 So I took the time to read through all the docs.  I can see how many times you remind us that the store needs subscribers.  Haha, my fault for missing all of them.  Does having a template bind to a selector count as a subscription?  
ie: ```{{ authService.user | async }}``` in template
